p2p/kademlia/dht.go

```diff
@@ -35,661 +39,436 @@ type Options struct {
+					log.WithContext(ctx).Errorf("iterate store data: %v", err)
+				}
+			}
 		}
-	}
 
-	return value, exists, nil
-}
+		return nil
+	})
 
-// NumNodes returns the total number of nodes stored in the local routing table
-func (dht *DHT) NumNodes() int {
-	return dht.ht.totalNodes()
+	return nil
 }
 
-// GetSelfID returns the base58 encoded identifier of the local node
-func (dht *DHT) GetSelfID() string {
-	str := b58.Encode(dht.ht.Self.ID)
-	return str
-}
+// Stop the distributed hash table
+func (s *DHT) Stop(ctx context.Context) {
+	if s.done != nil {
+		close(s.done)
+	}
 
-// GetNetworkAddr returns the publicly accessible IP and Port of the local
-// node
-func (dht *DHT) GetNetworkAddr() string {
-	return dht.networking.getNetworkAddr()
+	// stop the network
+	s.network.Stop(ctx)
 }
 
-// UseStun checks if DHT is using STUN.
-func (dht *DHT) UseStun() bool {
-	return dht.options.UseStun
+// a hash key for the data
+func (s *DHT) hashKey(data []byte) []byte {
+	sha := sha1.Sum(data)
+	return sha[:]
 }
 
-// CreateSocket attempts to open a UDP socket on the port provided to options
-func (dht *DHT) CreateSocket() error {
-	ip := dht.options.IP
-	port := dht.options.Port
+// Store the data into the network
+func (s *DHT) Store(ctx context.Context, data []byte) (string, error) {
+	key := s.hashKey(data)
 
-	netMsgInit()
-	dht.networking.init(dht.ht.Self)
+	// replicate time for the key
+	replication := time.Now().Add(defaultReplicateTime)
 
-	publicHost, publicPort, err := dht.networking.createSocket(ip, port, dht.options.UseStun, dht.options.StunAddr)
-	if err != nil {
-		return err
+	// store the key to local storage
+	if err := s.store.Store(ctx, key, data, replication); err != nil {
+		return "", fmt.Errorf("store data to local storage: %v", err)
 	}
 
-	if dht.options.UseStun {
-		dht.ht.setSelfAddr(publicHost, publicPort)
+	// iterative store the data
+	if _, err := s.iterate(ctx, IterateStore, key, data); err != nil {
+		return "", fmt.Errorf("iterative store data: %v", err)
 	}
 
-	return nil
+	return base58.Encode(key), nil
 }
 
-// Listen begins listening on the socket for incoming messages
-func (dht *DHT) Listen(ctx context.Context) error {
-	if !dht.networking.isInitialized() {
-		return errors.New("socket not created")
+// Retrieve data from the networking using key. Key is the base58 encoded
+// identifier of the data.
+func (s *DHT) Retrieve(ctx context.Context, key string) ([]byte, error) {
+	decoded := base58.Decode(key)
+	if len(decoded) != K {
+		return nil, fmt.Errorf("invalid key: %v", key)
 	}
 
-	group, _ := errgroup.WithContext(ctx)
-	group.Go(func() error {
-		return dht.listen(ctx)
-	})
-	group.Go(func() error {
-		return dht.timers(ctx)
-	})
-	group.Go(func() error {
-		return dht.networking.listen(ctx)
-	})
-	return group.Wait()
+	// retrieve the key/value from local storage
+	value, err := s.store.Retrieve(ctx, decoded)
+	if err != nil {
+		return nil, fmt.Errorf("store retrieve: %v", err)
```

Comment:
if error happen get from local storage(like data corruption), do you think should try get from another peer, to prevent user get stuck here?
certainly should have a way to monitor health of local storage.

Reply:
Good idea